### PR TITLE
JDK-8164094: javadoc allows to create a @link to a non-existent method

### PR DESCRIPTION
**Pull request by hns**
Please review a simple fix in `JavacTrees` to only look up member references in the enclosing type if the reference does not contain an explicit type name. For example, `@see #method()` in a a doc comment of class `Outer.Inner` should find method `Outer.method()`, but `@see Inner#method()` should not (regardless of comment location).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8325036](https://bugs.openjdk.org/browse/JDK-8325036) to be approved
- [x] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issues
 * [JDK-8164094](https://bugs.openjdk.org/browse/JDK-8164094): javadoc allows to create a @link to a non-existent method (**Bug** - P3)
 * [JDK-8325036](https://bugs.openjdk.org/browse/JDK-8325036): javadoc allows to create a @link to a non-existent method (**CSR**)

### Reviewers
 * [Pavel Rappo](https://openjdk.org/census#prappo) (@pavelrappo - **Reviewer**) ⚠️ Review applies to [4f08440e](https://git.openjdk.org/jdk/pull/17069/files/4f08440ef32e188b443a91dcc0153c828cb3842c)
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to [4f08440e](https://git.openjdk.org/jdk/pull/17069/files/4f08440ef32e188b443a91dcc0153c828cb3842c)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17069/head:pull/17069` \
`$ git checkout pull/17069`

Update a local copy of the PR: \
`$ git checkout pull/17069` \
`$ git pull https://git.openjdk.org/jdk.git pull/17069/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17069`

View PR using the GUI difftool: \
`$ git pr show -t 17069`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17069.diff">https://git.openjdk.org/jdk/pull/17069.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17069#issuecomment-1850910724)

### CODE REVIEW

**Comment by hns** [comment]
There are a few cases of this linking pattern in OpenJDK that need to be fixed before this can be integrated. I filed JDK-8321889.

**Review by pavelrappo — COMMENTED** [review]
It's nice to see this issue and its individual sightings being addressed.

While I'm a Reviewer, I'm not a compiler person, so my approval means little here. Wait for a compiler person to approve it. Perhaps they might also suggest a better name for the `explicitType` parameter.

Separately, it was interesting to see that a field is first searched in the enclosing class or interface, and if not found, searched in the superclass or a superinterface. The order in which a method is searched is the opposite: the superclass or a superinterface, then the enclosing class or interface.

**Comment by pavelrappo** [comment]
> individual sightings being addressed.

I meant in a separate PR #17078.

**Comment by jddarcy** [comment]
Should this issue have a CSR for the behavioral change?

**Comment by jonathan-gibbons** [comment]
> It's nice to see this issue and its individual sightings being addressed.
> 
> While I'm a Reviewer, I'm not a compiler person, so my approval means little here. Wait for a compiler person to approve it. Perhaps they might also suggest a better name for the `explicitType` parameter.
> 
> Separately, it was interesting to see that a field is first searched in the enclosing class or interface, and if not found, searched in the superclass or a superinterface. The order in which a method is searched is the opposite: the superclass or a superinterface, then the enclosing class or interface.

Maybe @lahodaj can help.

**Comment by hns** [comment]
> Should this issue have a CSR for the behavioral change?

My understanding of the issue is that the lookup of explicit-class member references in enclosing classes was an unintended consequence of doing the same for classless references from within nested classes. So I consider this a bug fix, not a feature change (even though it unfortunately changes javadoc to become stricter in what it accepts).

Actually, the feature of extending member reference lookup to enclosing classes is not specified in the taglet spec, which only says this:

> When the reference is to a member of the same class as that containing the documentation comment, all parts of the reference up to and including the # may be omitted, although the '#' may be retained for clarity.

This could be extended to include enclosing classes, but this is existing behaviour, not changed by this issue.

**Comment by pavelrappo** [comment]
> Separately, it was interesting to see that a field is first searched in the enclosing class or interface, and if not found, searched in the superclass or a superinterface. The order in which a method is searched is the opposite: the superclass or a superinterface, then the enclosing class or interface.

As I said before, it's a separate issue. However, I still wonder what the rationale for that difference is; it's not obvious to me. It also doesn't strike me as the right behaviour.

FWIW, as a local experiment I changed code so that both methods and fields were first searched in the superclass and superinterfaces and then in the enclosing class or interface. Out of the above two alternatives, this order seems more correct. All tier1 tests still succeeded.

**Comment by lahodaj** [comment]
For "normal" field or method (simple-name) lookup, the same-named field or method from the enclosing scope is shadowed by the declaration in the current ("nested") scope (JLS 6.4.1). The "nested" scope includes inherited fields and methods. So, even for fields, it might be preferable to first search supertypes, and only then enclosing scopes.

There is a caveat: the inheritance rules are a bit complex, like for example `private`, inaccessible or hidden(*) fields are not inherited (JLS 8.3); rules for methods are even a bit more complex (JLS 8.4.8). I.e. if we just change the search order, we may end up looking up a private field from a supertype, instead of the more correct field from an enclosing scope.

(*) hidden fields are not really an issue here.

There's also one more specialty for methods: if there's any method of name `m` in a given scope, even if it is not applicable due to parameter types, the search ends in that scope, and does not go to the enclosing scopes (a method named `m` shadows all methods named `m` with any parameters in the enclosing scope). (JLS 6.4.1)

I am not sure how strictly correct javadoc needs to be, but if we want to get closer to the normal "simple-name" lookup, we would probably need to tighten up the code.

I was briefly looking whether we could reuse the standard lookup - might be doable for fields, but not sure about methods. Would need some more investigations, I am afraid.

**Comment by pavelrappo** [comment]
@hns would you be able to create a linked bug to track that lookup order issue?

**Review by pavelrappo — APPROVED** [review]

**Comment by jonathan-gibbons** [comment]
I note the preceding discussion including @pavelrappo's notes about order of lookup, and the separate issue to pursue this further.

It's also not clear to me that the statement in the description is "right":  _but @see Inner#method() should not_ 
If `Inner.method()` has a meaning (making allowances for `static` or not, etc) in nearby code, then shouldn't `Inner#method()` have a similar meaning?

I think that we should
1.  determine the lexical scope for the doc comment
2. follow the JLS rules as much as possible, with matching examples for doc comments and Java code as much as possible
3. clearly specify that doc comment names follow the JLS rules, except when they're not, and clearly specify any exceptions to those rules

All that is out of scope for here, and can be handled in [JDK-8324320](https://bugs.openjdk.org/browse/JDK-8324320), so I have one last question before I add my approval (or Pavel can lower the `/reviewer` count):
Does this change affect the published/latest JDK API documentation in any way?  I'm guessing not, but we've been surprised before.

**Comment by hns** [comment]
> All that is out of scope for here, and can be handled in [JDK-8324320](https://bugs.openjdk.org/browse/JDK-8324320), so I have one last question before I add my approval (or Pavel can lower the `/reviewer` count): Does this change affect the published/latest JDK API documentation in any way? I'm guessing not, but we've been surprised before.

It affects the published JDK API docs insofar as it exposed [3 cases of method references with wrong types](https://bugs.openjdk.org/browse/JDK-8321889). This has already been fixed back in December.

**Comment by hns** [comment]
Just to make sure I generated JDK API docs with and without this fix and there was no diff in outputs.

**Comment by jddarcy** [comment]
> > Should this issue have a CSR for the behavioral change?
> 
> My understanding of the issue is that the lookup of explicit-class member references in enclosing classes was an unintended consequence of doing the same for classless references from within nested classes. So I consider this a bug fix, not a feature change (even though it unfortunately changes javadoc to become stricter in what it accepts).
> 
> Actually, the feature of extending member reference lookup to enclosing classes is not specified in the taglet spec, which only says this:
> 
> > When the reference is to a member of the same class as that containing the documentation comment, all parts of the reference up to and including the # may be omitted, although the '#' may be retained for clarity.
> 
> This could be extended to include enclosing classes, but this is existing behaviour, not changed by this issue.

Looping back on this, independent of what the intended or specified behavior was before, if the behavior is changing now, even in agreed upon ways, the change should probably go through CSR review.

**Comment by hns** [comment]
> Looping back on this, independent of what the intended or specified behavior was before, if the behavior is changing now, even in agreed upon ways, the change should probably go through CSR review.

Ok, I've written a CSR.

File: test/langtools/tools/javac/doctree/ReferenceTest.java
Comment (jonathan-gibbons) [inline test/langtools/tools/javac/doctree/ReferenceTest.java:2]: 2024?

**Review by jonathan-gibbons — APPROVED** [review]

File: test/langtools/tools/javac/doctree/ReferenceTest.java
Comment (hns) [reply test/langtools/tools/javac/doctree/ReferenceTest.java]: Updated

**Pull request closed without merging**
